### ISSUE TITLE
SAW: initDefaultSprite_Contract's character pointers are confusing

### ISSUE THREAD

**Issue by ddfreeman** [weaversa/cryptol-course#216]
The current implementation of initDefaultSprite_Contract creates two symbolic pointers for characters - `character`, for the actual character used as input to initDefaultSprite, and `tempCharacter_p`, which seems to be whatever `sprite_p->character` is before reassignment.

This seems to be confusing, since it's not clear at first why there are two characters (or, at least, what _appears_  to be two characters) being allocated. It might be better to, instead of assuming that `sprite_p->character` points to some garbage address represented by `tempCharacter_p`, require that it points to NULL. This would both reflect good C coding practice and give a practical example of using `null()`.

**Comment by djmorel** [comment]
You bring up some fair points, and are spot on as to why there are two characters! The contract's `character` symbolic pointer represents the function's `character` input argument, whereas `tempCharacter_p` is the initial value for the `sprite->character` field.

Given that we don't assert any preconditions for the `sprite` input argument, we could instead just create `sprite_p` via the following:
```python
sprite_p = self.alloc(alias_ty("struct.sprite_t"))
```
Doing so would allocate all of the memory we need for our `sprite` input argument, and remove the need for `tempCharacter_p`, `ty`, `frames`, `xPos`, and `yPos`. It would also improve our proof's performance.

The contract was configured in this way to show an alternative method to create structs and open the door for preconditions. The lab's documentation could definitely be a bit clearer on this setup, and actually take advantage of the configuration to show preconditions for struct fields.

I like your idea about asserting the precondition that `tempCharacter_p` points to NULL. I'll update the example to reflect that :)

**Comment by weaversa** [comment]
It's a tad hard to tell what the right thing is because we don't see a call to `initDefaultSprite`. Should the caller set the value be NULL? I kind of assume the way this function would be called is for the caller to `malloc` the struct and then immediately pass it into this function. Asking the caller to remember to set some values to NULL will cause some issues.

I can see @djmorel's stated benefit of the example, but if I were to implement such a function in C, I would perform the allocation of the `sprite` inside this function, initialize it as is done now, and return a pointer to the newly malloc'd value, rather than trusting the caller to perform the allocation. Same for `initDefaultPlayer`. 

In my mind it should look something like:

```
sprite_t *createDefaultSprite(character_t* character)
{
  sprite_t *sprite = (sprite_t *)malloc(1 * sizeof(sprite_t));

  // Initialize the character to the passed pointer
  sprite->character = character;

  // Initialize the sprite frames to the default asset
  for (uint8_t i = 0; i < GAITS; i++)
  {
    for (uint8_t j = 0; j < DIRECTIONS; j++)
    {
      for (uint8_t k = 0; k < ANIMATION_STEPS; k++)
      {
        sprite->frames[i][j][k] = 0x00;
      }
    }
  }

  // Initialize sprite's default position
  sprite->xPos = 1;
  sprite->yPos = 2;

  return sprite;
}
```

**Comment by djmorel** [comment]
I agree that `initDefaultSprite` is a bit ambiguous in its current state, @weaversa. When I first wrote the function, I envisioned `initDefaultSprite`'s caller would create the `sprite_t` object and then pass the reference to `initDefaultSprite` like so:
```c
uint32_t spriteCaller()
{
  uint32_t status = FAILURE;
  sprite_t sprite;
  character_t character;
  // Assume some useful setup happens for character

  status = initDefaultSprite(&sprite, &character);
  if (status != SUCCESS)
  {
    // Do something
  }

  return status;
}
```

With that said, I like how `createDefaultSprite`  is an example of using `malloc` and returning a struct. That setup would spice up the other examples, which currently just return either an `uint32_t` or nothing at all.

As for reworking the example to better represent explicit structs, the setup could be reconfigured to check for preconditions on the `character` input argument. Let's say that there is a rule where a character may only be assigned to one `sprite_t` object. To satisfy this assumption, a `boolean` field could be added to the `character_t` struct that flags whether or not a `sprite_t` object uses the character. Sure we could just use a pointer instead, but then having `sprite` and `character` pointing to each other seems silly. 

It probably would be better to rework the relationship between `character_t` and `sprite_t`, and then have the explicit struct setup check for that relationship in a precondition.

Based on how this goes, I'll update the setup for `initDefaultPlayer` as well.
